evm/vm_state.py

```diff
@@ -0,0 +1,325 @@
+        state.db = None
+        state._trie = None
+
+    #
+    # Snapshot and Revert
+    #
+    def snapshot(self):
+        """
+        Perform a full snapshot of the current state.
+
+        Snapshots are a combination of the state_root at the time of the
+        snapshot and the checkpoint_id returned from the journaled DB.
+        """
+        return (self.block_header.state_root, self._chaindb.snapshot())
+
+    def revert(self, snapshot):
+        """
+        Revert the VM to the state at the snapshot
+        """
+        state_root, checkpoint_id = snapshot
+
+        with self.state_db() as state_db:
+            # first revert the database state root.
+            state_db.root_hash = state_root
+            # now roll the underlying database back
+
+        self._chaindb.revert(checkpoint_id)
+
+    def commit(self, snapshot):
+        """
+        Commits the journal to the point where the snapshot was taken.  This
+        will destroy any journal checkpoints *after* the snapshot checkpoint.
+        """
+        _, checkpoint_id = snapshot
+        self._chaindb.commit(checkpoint_id)
+
+    #
+    # Access self.prev_headers (Read-only)
+    #
+    def get_ancestor_hash(self, block_number):
+        """
+        Return the hash for the ancestor with the given block number.
+        """
+        ancestor_depth = self.block_header.block_number - block_number
+        if ancestor_depth > MAX_PREV_HEADER_DEPTH or ancestor_depth < 1:
+            return b''
+        header = self.get_block_header_by_hash(self.block_header.parent_hash)
+        while header.block_number != block_number:
+            header = self.get_block_header_by_hash(header.parent_hash)
+        return header.hash
+
+    def get_block_header_by_hash(self, block_hash):
+        """
+        Returns the block header by hash.
+        """
+        for value in self.prev_headers:
+            if value.hash == block_hash:
+                return value
+        raise BlockNotFound(
+            "No block header with hash {0} found in self.perv_headers".format(
+                encode_hex(block_hash),
+            )
+        )
+
+    def get_parent_header(self, block_header):
+        """
+        Returns the header for the parent block.
+        """
+        return self.get_block_header_by_hash(block_header.parent_hash)
+
+    def is_key_exists(self, key):
+        """
+        Check if the given key exsits in chaindb
+        """
+        return self._chaindb.exists(key)
+
+    #
+    # Computation
+    #
+    def get_computation(self, message):
+        """Return state object
+        """
+        if self.computation_class is None:
+            raise AttributeError("No `computation_class` has been set for this VMState")
+        else:
+            computation = self.computation_class(self, message)
+        return computation
+
+    #
+    # Execution
+    #
+    @classmethod
+    def apply_transaction(
+            cls,
+            vm_state,
```

Comment:
It feels weird to have a `BaseVMState` method that takes a `vm_state` argument, but maybe I'm missing something?